app/src/main/java/ch/hikemate/app/model/elevation/ElevationServiceRepository.kt

```diff
@@ -1,113 +1,265 @@
 package ch.hikemate.app.model.elevation
 
+import android.util.Log
 import ch.hikemate.app.model.route.LatLong
 import java.io.IOException
+import java.util.Date
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
 import kotlinx.serialization.Serializable
 import kotlinx.serialization.json.Json.Default.decodeFromString
+import kotlinx.serialization.json.add
 import kotlinx.serialization.json.buildJsonArray
-import kotlinx.serialization.json.buildJsonObject
-import kotlinx.serialization.json.put
-import okhttp3.MediaType.Companion.toMediaType
 import okhttp3.OkHttpClient
-import okhttp3.RequestBody
-import okhttp3.RequestBody.Companion.toRequestBody
-
-private const val BASE_URL = "https://api.open-elevation.com/api/v1/lookup"
-private const val LOCATION_KEY = "locations"
-private const val LATITUDE_KEY = "latitude"
-private const val LONGITUDE_KEY = "longitude"
-private const val ACCEPT_HEADER = "Accept"
-private const val APPLICATION_JSON = "application/json"
-private const val CONTENT_TYPE_HEADER = "Content-Type"
-private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
 
-/** Serializable data class for the response from the Elevation API */
-@Serializable data class ElevationResponse(val results: List<ElevationResult>)
+private const val BASE_URL = "https://www.elevation-api.eu/v1/elevation?pts="
+
+// We need to wait some time before updating the cache to allow for multiple requests to be
+// batched together
+private const val CACHE_UPDATE_DELAY = 500L
+
+// Limit the cache size to avoid memory issues
+// This has been tested with a cache size of 100000, on a Google Pixel 8 Pro, without any issues
+private const val CACHE_MAX_SIZE = 150000
+
+// The maximum number of failed requests before giving up
+private const val MAX_FAILED_REQUESTS = 5
+
+// Delay to add per failed request
+private const val FAILED_REQUEST_DELAY = 1000L
+
+// The max number of coordinates that can be sent in a single request
+private const val MAX_COORDINATES_PER_REQUEST = 500
 
 /** Serializable data class for the response from the Elevation API */
-@Serializable
-data class ElevationResult(val latitude: Double, val longitude: Double, val elevation: Double)
+@Serializable data class ElevationResponse(val elevations: List<Double?>)
```

Comment:
It's simply determined by the API! The API can return null as a value, and since this is a Serializable class, it needs to allow it. I'll add some documentation about it!